controllers/nstemplateset/space_roles.go

```diff
@@ -61,33 +62,36 @@ func (r *spaceRolesManager) ensure(logger logr.Logger, nsTmplSet *toolchainv1alp
 			toolchainv1alpha1.ProviderLabelKey: toolchainv1alpha1.ProviderLabelValue,
 			toolchainv1alpha1.OwnerLabelKey:    nsTmplSet.GetName(),
 		}
-		logger.Info("creating space role objects")
+		logger.Info("creating space role objects", "count", len(spaceRoleObjs))
 		// create (or update existing) objects based the tier template
 		if _, err = r.ApplyToolchainObjects(logger, spaceRoleObjs, labels); err != nil {
 			return false, r.wrapErrorWithStatusUpdate(logger, nsTmplSet, r.setStatusNamespaceProvisionFailed, err, "failed to provision namespace '%s' with space roles", ns.Name)
 		}
 
-		if err = deleteObsoleteObjects(logger, r.Client, lastAppliedSpaceRoleObjs, spaceRoleObjs); err != nil {
+		if err := deleteObsoleteObjects(logger, r.Client, lastAppliedSpaceRoleObjs, spaceRoleObjs); err != nil {
 			return false, r.wrapErrorWithStatusUpdate(logger, nsTmplSet, r.setStatusUpdateFailed, err, "failed to delete redundant objects in namespace '%s'", ns.Name)
 		}
 
-		// store the space roles in an annotation at the namespace level, so we know what was applied and how to deal with
-		// diffs when the space roles are changed (users added or removed, etc.)
-		sr, err := json.Marshal(nsTmplSet.Spec.SpaceRoles)
-		if err != nil {
-			return false, r.wrapErrorWithStatusUpdate(logger, nsTmplSet, r.setStatusProvisionFailed, err,
-				"failed update namespace annotation")
-		}
-		if ns.Annotations == nil {
-			ns.Annotations = map[string]string{}
-		}
-		ns.Annotations[toolchainv1alpha1.LastAppliedSpaceRolesAnnotationKey] = string(sr)
-		if err := r.Client.Update(context.TODO(), &ns); err != nil { // nolint:gosec
-			return false, r.wrapErrorWithStatusUpdate(logger, nsTmplSet, r.setStatusProvisionFailed, err,
-				"failed update namespace annotation")
+		if !reflect.DeepEqual(nsTmplSet.Spec.SpaceRoles, lastAppliedSpaceRoles) {
+			// store the space roles in an annotation at the namespace level, so we know what was applied and how to deal with
+			// diffs when the space roles are changed (users added or removed, etc.)
+			sr, err := json.Marshal(nsTmplSet.Spec.SpaceRoles)
+			if err != nil {
+				return false, r.wrapErrorWithStatusUpdate(logger, nsTmplSet, r.setStatusProvisionFailed, err,
+					fmt.Sprintf("failed to marshal space roles to update '%s' annotation on namespace", toolchainv1alpha1.LastAppliedSpaceRolesAnnotationKey))
+			}
+			if ns.Annotations == nil {
+				ns.Annotations = map[string]string{}
+			}
+			ns.Annotations[toolchainv1alpha1.LastAppliedSpaceRolesAnnotationKey] = string(sr)
+			logger.Info("updated annotation on namespace", toolchainv1alpha1.LastAppliedSpaceRolesAnnotationKey, string(sr))
```

Comment:
It looks like this message got moved before the call to Update when it was previously after? I guess it should be moved to after the Update?

Reply:
good point. Reverted in ef1c026